src/main/java/hudson/remoting/Launcher.java

```diff
@@ -456,25 +480,146 @@ private SSLSocketFactory getSSLSocketFactory()
+        }
+        if (proxyCredentials != null) {
+            jnlpArgs.add("-proxyCredentials");
+            jnlpArgs.add(proxyCredentials);
+        }
+        if (noKeepAlive) {
+            jnlpArgs.add("-noKeepAlive");
+        }
+        if (workDir != null) {
+            jnlpArgs.add("-workDir");
+            jnlpArgs.add(workDir.getAbsolutePath());
+            jnlpArgs.add("-internalDir");
+            jnlpArgs.add(internalDir);
+            if (failIfWorkDirIsMissing) {
+                jnlpArgs.add("-failIfWorkDirIsMissing");
+            }
+        }
+        if (candidateCertificates != null && !candidateCertificates.isEmpty()) {
+            for (String c : candidateCertificates) {
+                jnlpArgs.add("-cert");
+                jnlpArgs.add(c);
+            }
+        }
+        if (noCertificateCheck) {
+            jnlpArgs.add("-noCertificateCheck");
+        }
+
+        // Parse the server's pseudo-JNLP output
+        Launcher bootstrap = new Launcher();
+        CmdLineParser parser = new CmdLineParser(bootstrap);
+        parser.parseArgument(jnlpArgs.toArray(new String[0]));
+        normalizeArguments(bootstrap);
+        validateInboundAgentArgs(bootstrap);
+
+        // Apply the results
+        assert urls.isEmpty();
+        urls.addAll(bootstrap.urls);
+        if (bootstrap.secret != null) {
+            secret = bootstrap.secret;
+        }
+        if (bootstrap.name != null) {
+            name = bootstrap.name;
+        }
+        if (bootstrap.webSocket) {
+            webSocket = true;
+        }
+        if (bootstrap.tunnel != null) {
+            tunnel = bootstrap.tunnel;
+        }
+        if (bootstrap.workDir != null) {
+            workDir = bootstrap.workDir;
+        }
+        if (!WorkDirManager.DirType.INTERNAL_DIR.getDefaultLocation().equals(bootstrap.internalDir)) {
+            internalDir = bootstrap.internalDir;
+        }
+        if (bootstrap.failIfWorkDirIsMissing != WorkDirManager.DEFAULT_FAIL_IF_WORKDIR_IS_MISSING) {
+            failIfWorkDirIsMissing = bootstrap.failIfWorkDirIsMissing;
+        }
+    }
+
+    private static void normalizeArguments(Launcher launcher) throws CmdLineException {
+        if (!launcher.args.isEmpty()) {
+            if (launcher.args.size() != 2) {
+                throw new CmdLineException(null, "Two arguments required, but got " + launcher.args);
+            }
+            if (launcher.secret == null) {
+                launcher.secret = launcher.args.get(0);
+            } else {
+                throw new CmdLineException(null, "Cannot provide secret via both named and positional arguments");
+            }
+            if (launcher.name == null) {
+                launcher.name = launcher.args.get(1);
+            } else {
+                throw new CmdLineException(null, "Cannot provide name via both named and positional arguments");
+            }
+            launcher.args.clear();
+        }
+    }
```

Comment:
This is here to retain backwards compatibility with the positional argument calling convention. It converts the positional arguments, if present, to their named argument replacements.